Add ARP collector for Linux

skottler opened the pull request:
This introduces a simple ARP collector with initial support for Linux. It parses `/proc/net/arp` and emits a count of ARP entries on a per-device basis.

/cc https://github.com/prometheus/node_exporter/issues/535

grobie commented on collector/arp_linux.go:
```diff
@@ -0,0 +1,84 @@
+	scanner := bufio.NewScanner(data)
+	entries := make(map[string]uint32)
+
+	for scanner.Scan() {
```
`scanner.Scan()` can fail. You need to check `scanner.Err()` and handle that.

grobie commented on collector/arp_linux.go:
```diff
@@ -0,0 +1,84 @@
+	return parseArpEntries(file), nil
+}
+
+func parseArpEntries(data *os.File) map[string]uint32 {
```
The only requirement is to pass something implementing the `io.Reader` interface, so I'd reflect that in the function signature.

grobie commented on collector/arp_linux.go:
```diff
@@ -0,0 +1,84 @@
+	entries := make(map[string]uint32)
+
+	for scanner.Scan() {
+		columns := strings.Split(string(scanner.Text()), " ")
```
`scanner.Text()` returns a string. The type conversion is unnecessary.

grobie replied:
Also, check `strings.Fields()`

skottler commented:
Thanks for the insanely quick feedback @grobie! :bowing_man: I've adopted the changes you suggested.

> The consensus has been to handle procfs parsing in https://github.com/prometheus/procfs and use that implementation here though.

This PR implements only a tiny portion of what potentially needs to get parsed from `/proc/arp/net` as it only covers what's needed to expose the count per device. Is it okay to add a very much incomplete API to the procfs package or would you prefer that more complete parsing support gets implemented first?

grobie commented:
You're lucky I had to push my hiking trip back by one day. Now I'm bored in
Patagonia and have time to review PRs ;-)

I'm fine with either merging the current state with a comment saying
something along the lines of "if more /proc/net/arp parsing support is
needed, move this to prometheus/procfs" or by adding full parsing support
to procfs now. Shouldn't be to hard to parse the whole line and return a
slice like []ARPEntry.

On Sun, Apr 2, 2017 at 4:39 PM Sam Kottler <notifications@github.com> wrote:

> Thanks for the insanely quick feedback @grobie <https://github.com/grobie>!
> 🙇 I've adopted the changes you suggested.
>
> The consensus has been to handle procfs parsing in
> https://github.com/prometheus/procfs and use that implementation here
> though.
>
> This PR implements only a tiny portion of what potentially needs to get
> parsed from /proc/arp/net as it only covers what's needed to expose the
> count per device. Is it okay to add a very much incomplete API to the
> procfs package or would you prefer that more complete parsing support gets
> implemented first?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/node_exporter/pull/540#issuecomment-291009575>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAANaGCHJLZgjnK7MJcKmrMSgmEGoFuHks5rr_lvgaJpZM4Mw4_K>
> .
>

grobie commented on collector/arp_linux.go:
```diff
@@ -0,0 +1,94 @@
+	entries := make(map[string]uint32)
+
+	for scanner.Scan() {
+		if err := scanner.Err(); err != nil {
```
`Err()` will always be nil if `Scan()` is not nil. Check the example in https://golang.org/pkg/bufio/#example_Scanner_custom

skottler commented:
> I'm fine with either merging the current state with a comment saying something along the lines of "if more /proc/net/arp parsing support is needed, move this to prometheus/procfs" or by adding full parsing support to procfs now.

I added a comment in that vain to the current parsing function. If you're comfortable merging this in its current state I'm happy to follow up with `/proc/net/arp` support in the procfs package and then switching `node_exporter` to use that. Doing that now or after merge are both fine with me.

skottler commented:
```
# netbsd-armv5

>> building binaries

 >   node_exporter

# github.com/prometheus/node_exporter

/usr/local/go/pkg/tool/linux_amd64/link: running arm-linux-gnueabi-gcc failed: exit status 1

/tmp/go-link-863345215/000001.o: In function `_cgo_41b2da71c790_C2func_getaddrinfo':

/usr/local/go/src/net/cgo-gcc-prolog:45: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking

/usr/lib/gcc-cross/arm-linux-gnueabi/6/../../../../arm-linux-gnueabi/lib/crt1.o: In function `_start':

(.text+0x34): undefined reference to `main'

collect2: error: ld returned 1 exit status
```

The tests look to be failing on netbsd with ARMv5. I hope that's a CI or build issue rather than actual failure because debugging that is gonna be...challenging 😂

juliusv commented on collector/arp_linux.go:
```diff
@@ -0,0 +1,98 @@
+	Factories["arp"] = NewArpCollector
+}
+
+// NewArpCollector returns a new Collector exposing ARP stats.
```
Arp -> ARP (https://github.com/golang/go/wiki/CodeReviewComments#initialisms)

Same below.

juliusv commented on collector/arp_linux.go:
```diff
@@ -0,0 +1,98 @@
+	return &arpCollector{
+		count: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "arp",
+				"count"), "ARP entries by device",
```
This is kind of a confusing indentation/line break combination. If a function call is broken into multiple lines, I wouldn't then expect some other argument to follow it on the same line...

How about:

```go
count: prometheus.NewDesc(
    prometheus.BuildFQName(Namespace, "arp", "count"),
    "ARP entries by device",
    []string{"device"}, nil,
),
```

juliusv commented on collector/arp_linux.go:
```diff
@@ -0,0 +1,98 @@
+	return entries, nil
+}
+
+// This should get extracted to the github.com/prometheus/procfs package to
```
Add a `TODO: ` to the beginning of this.

juliusv commented:
@skottler Regarding the test errors, I don't see how you could be causing them, given that you are not introducing any architecture-specific functionality, or even cgo stuff...

mdlayher commented on collector/arp_linux.go:
```diff
@@ -0,0 +1,98 @@
+	for scanner.Scan() {
+		columns := strings.Fields(scanner.Text())
+
+		if columns[0] != "IP" {
```
This will panic if `strings.Fields` returns an empty slice, so it's probably worth adding a bounds check too.

https://play.golang.org/p/09sEiI2gEN

:wave: Hey Sam!

juliusv replied:
Yes, always bounds check first!

grobie commented on collector/arp_linux.go:
```diff
@@ -0,0 +1,98 @@
+func NewARPCollector() (Collector, error) {
+	return &arpCollector{
+		count: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "arp", "count"),
```
We don't include things like `count` in gauge metric names. I'd suggest `node_arp_entries` or maybe `node_arp_hosts`?

skottler commented:
@juliusv I've addressed your feedback in 43c748000c6f70f3d38132e973d38f39eb26b8cf, c1e6b6a764fc53419b0a2a23a80b444be7980e36, and eb3934a5208168e963c3dccf82591e086c660a6d.

@mdlayher :wave: hope you're doing well! I addressed your comment in b5ba1021250571b6e0dd9eaeca511a3547a4d9fc.

grobie commented on collector/arp_linux.go:
```diff
@@ -0,0 +1,100 @@
+	for scanner.Scan() {
+		columns := strings.Fields(scanner.Text())
+
+		if len(columns) > 0 {
```
Sorry for all the nitpicking. Idiomatic go tries to reduce the amount of nesting as much as possible, so `if len(columns) == 0 { continue }` would be more common here. Though, I think we should actually return an error here, such line would be completely unexpected and tell us about either a broken or changed ARP file. So I suggest:

```go
if len(columns) < 6 {
    return nil, fmt.Errorf("unexpected ARP table file")
}

// ...
```

grobie commented on collector/arp_linux.go:
```diff
@@ -0,0 +1,100 @@
+// to support more complete parsing of /proc/net/arp. Instead of adding
+// more fields to this function's return values it should get moved and
+// changed to support each field.
+func parseArpEntries(data io.Reader) (map[string]uint32, error) {
```
s/Arp/ARP/

grobie commented on collector/arp_linux.go:
```diff
@@ -0,0 +1,100 @@
+	}, nil
+}
+
+func getArpEntries() (map[string]uint32, error) {
```
s/Arp/ARP/

grobie commented on collector/arp_linux.go:
```diff
@@ -0,0 +1,100 @@
+		return fmt.Errorf("could not get ARP entries: %s", err)
+	}
+
+	for device, entryCount := range entries {
```
As a note, it's common in go to use single letter variables in such short loops. I'm fine with the descriptive names though.

skottler commented:
@grobie thanks! I've updated this including your feedback in 532677e35af3fe9cdc9885a6708bd44106bd14d9.

mdlayher commented:
Seems like this would be a reasonable collector to enable by default as well. Thoughts on that from anyone?

grobie commented:
:+1:

Please also add it to the readme.

On Sun, Apr 2, 2017, 21:19 Matt Layher <notifications@github.com> wrote:

> Seems like this would be a reasonable collector to enable by default as
> well. Thoughts on that from anyone?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/node_exporter/pull/540#issuecomment-291025655>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAANaOGIDJtC4cUvesSdBasE9llh7T1_ks5rsDsWgaJpZM4Mw4_K>
> .
>

skottler commented:
> Seems like this would be a reasonable collector to enable by default as well.

Done in 5ebb24edfb0666f2cf5651dfdaea6622bd48f152.

> Please also add it to the readme.

Done in 2a82e3d4e1888b706bf3281b162ef87e86b3be96.

grobie commented:
I believe the test failures are the result of @juliusv latest change to build Prometheus components with go1.8 by default. Maybe our build yoda @sdurrheimer can help here? :-)

grobie commented:
Build issues fixed in master.
